### PR TITLE
Updating params

### PR DESCRIPTION
**Pull request by matthieutrs**
We propose to change the organisation of the algorithms' parameters as well as the prior. Each are now implemented as dictionaries, which has 2 benefits:
1. Overall simplified syntax: no more long lists of arguments for the Optim class functions;
2. No more dependency of the fixed-point operator to the prior, removing the dependency of the iterator to the iteration number when using unfolded architectures.

New syntax can be summarised as such:
```
prior = {'prox_g': Denoiser(model_spec)}  # define the prior (here, this is a PnP with a denoiser for the prox)
params_algo={'stepsize': [1.]*max_iter, 'g_param':  [0.01]*max_iter, 'lambda' : 1.}   # define the algorithm related parameters (e.g. stepsize, regularisation parameter...)
trainable_params = ['stepsize', 'g_param']   # Select trainable parameters
model = Unfolded(algo_name, params_algo=params_algo, trainable_params=trainable_params, 
                        data_fidelity=data_fidelity, max_iter=max_iter, prior=prior)  # Define the model
```

## 1. Removed dependency of the iterator to the iterates
**Old syntax:**
```
class FixedPoint(nn.Module):
    ...
    def forward(self, x, *args, **kwargs):
        ...
        for it in range(self.max_iter):
        
             x = self.iterator(x, it, *args)
```
Problem: the iterator explicitely depends on the iteration number.

Under the hood, the dependency to the iteration number mainly fuels the parameter selection. The new syntax makes the iterator dependent on parameters (such as e.g. a stepsize) which can be updated in the iteration loop.
**New syntax:**
```
class FixedPoint(nn.Module):
    ...
    def forward(self, x, *args, **kwargs):
        ...
        for it in range(self.max_iter):

              cur_prior = self.update_prior_fn(it)
              cur_params = self.update_params_fn_pre(it, x, x_prev) 
              x = self.iterator(x, cur_prior, cur_params, *args, **kwargs)
```

## 2. New parameter syntax
**Old syntax:**
```
class gStep(nn.Module):
    def __init__(self, stepsize=[1.] * 50, ...):
        super(gStep, self).__init__()
        self.stepsize = stepsize
        ...
     
    def forward(self, x, it):
        pass
```
Problem 1: `gStep` depends on the iteration number `it`.
Problem 2: the stepsize is tied to gStep.

Instead, we propose to have parameters like `stepsize` that are externally fed to gStep at the level of the `forward` method (solving problem 2). The (potential) dependency of the stepsize to the iteration number is dealt with at the level of the optimisation algorithm. 

**New syntax:**
```
class gStep(nn.Module):
    def __init__(self, ..., **kwargs):
        super(gStep, self).__init__()

        def forward(self, x, cur_params):
            pass
```

## 3. Definition of the prior
**Old syntax:** 
```
prox_g = ProxDenoiser(model_spec, sigma_denoiser=sigmas, stepsize=rhos, max_iter=max_iter)
model = Optim('HQS', prox_g=prox_g, data_fidelity=data_fidelity, stepsize=prox_g.stepsize,
             g_param=prox_g.sigma_denoiser, max_iter=max_iter)
```
Problem 1: many unnecessary parameters in the ProxDenoiser class (e.g the denoiser should not depend on `max_iter`; the denoiser threshold `sigma_denoiser` should not be a parameter of the denoiser but rather an input parameter to the forward function)
Problem 2: the Optim class gets a `prox_g` element assigned to the proposed model above. This is fine, but this means that the [Optim class also has other parameters](https://github.com/deepinv/deepinv/blob/main/deepinv/optim/optimizers.py#L55-L57), such as `grad_g`, assigned to None under the hood. Ideally, we would not want to have to define parameters that are not used.
_Problem 3:_ (solved from the above modification) A prox_g parameter (`sigma_denoiser`) is assigned as a parameter of the optim class (`g_param`), duplicating the parameter.

**New syntax:**
```
params_algo={'stepsize': stepsize, 'g_param': sigma_denoiser, 'lambda' : lamb}
prior = {'prox_g': Denoiser(model_spec)}
model = Optim(algo_name = 'HQS', prior=prior, data_fidelity=data_fidelity, early_stop=early_stop,
              max_iter=max_iter, params_algo=params_algo)
```

**Side updates:**
The prior related parameters (e.g. whether we use a replacement for a proximal operator, a score denoiser or else) are still used in the `OptimIterator` classes (think of PGD with either [a gradient step denoiser or with a proximity operator](https://github.com/deepinv/deepinv/blob/main/deepinv/optim/optim_iterators/pgd.py#L39-L41)).
The syntax moves from
```
self.prox_g(x, self.g_param[it], it)
```
to (see [here](https://github.com/deepinv/deepinv/blob/updating_params/deepinv/optim/optim_iterators/pgd.py#L42-L44))
```
 prior['prox_g'](x, cur_params['g_param'])
```

## 4. Denoiser nomenclature:
**Old syntax:**
```
class Denoiser(nn.Module):
    def __init__(self, model_spec=None, init=None, stepsize=1., sigma_denoiser=0.05, max_iter=None,
                 device=torch.device('cpu')):
        super(Denoiser, self).__init__()

       self.denoiser = make(model_spec)
       self.max_iter = max_iter
       self.device = device
       self.init = init
       self.sigma_denoiser = sigma_denoiser
       self.stepsize = stepsize

    def forward(self, x, sigma):
        return self.denoiser(x, sigma)
 ```
Problem: Unnecessary parameters. In particular, `self.sigma_denoiser` was used in other models (score denoising).
Old classes were:
- `ProxDenoiser()` for the usual denoiser in PnP
- `ScoreDenoiser()` for RED and diffusion

The new nomenclature removes any additional parameters to the denoiser. The `Denoiser` class is now replacing the previous `ProxDenoiser` class that had no added value. 
**New syntax:**
```
class Denoiser(nn.Module):

    def __init__(self, model_spec=None):
        super(Denoiser, self).__init__()
        self.denoiser = make(model_spec)

    def forward(self, x, sigma):
        return self.denoiser(x, sigma)
```

Denoising classes are now:
- `Denoiser()` for the usual denoiser in PnP
- `REDDenoiser()` for the RED approach (Identity minus denoiser)
- `ScoreDenoiser()` for diffusion (embedded Tweedie identity)

## 5. Warning points
1. Docs have not yet been updated accordingly.
2. Maybe it would be good to double check the sampling algorithms.

### CODE REVIEW

**Comment by tachella** [comment]
Great work

I'll wait to incorporate the changes that we discussed in Slack

- Remove REDDenoiser
- g_param -> sigma_denoiser
- clean up demos.
- prior=torch.nn.module?
